text/2978-stack_based_vec.md

```diff
@@ -0,0 +1,301 @@
+
+```rust
+fn main() {
+    // `array_vec` has a pre-allocated memory of 2048 bits that can store up to 64 decimals.
+    let mut array_vec: ArrayVec<i32, 64> = ArrayVec::new();
+
+    // Although reserved, there isn't anything explicitly stored yet
+    assert_eq!(array_vec.len(), 0);
+
+    // Initializes the first 32 bits with a simple '1' decimal or
+    // 00000000 00000000 00000000 00000001 bits
+    array_vec.push(1);
+
+    // Our vector memory is now split into a 32/2016 pair of initialized and
+    // uninitialized memory respectively
+    assert_eq!(array_vec.len(), 1);
+}
+```
+
+Of course, fixed buffers lead to some inflexibility because unlike `Vec`, the underlying capacity can not expand at run-time and there will never be more than 64 elements in the above example.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+`ArrayVec` is a contiguous memory block where elements can be collected, threfore, a collection by definition and even though `core::collections` doesn't exist, it is the most natural module placement.
+
+The API basically mimics most of the current `Vec` surface with some tweaks and additional methods to manage capacity. Notably, these additional methods are fallible versions of some well-known functions like `insert` that will return `Result` instead of panicking at run-time.
+
+```rust
+// core::collections
+
+pub struct ArrayVec<T, const N: usize> {
+    data: MaybeUninit<[T; N]>,
+    len: usize
+}
+
+impl<T, const N: usize> ArrayVec<T, N> {
+    // Constructors
+
+    pub unsafe fn from_raw_parts(_ptr: *mut T, _len: usize) -> Self;
+
+    #[inline]
+    pub const fn new() -> Self;
+
+    // Methods
+
+    #[inline]
+    pub fn as_mut_ptr(&mut self) -> *mut T;
+
+    #[inline]
+    pub fn as_mut_slice(&mut self) -> &mut [T];
+
+    #[inline]
+    pub fn as_ptr(&self) -> *const T;
+
+    #[inline]
+    pub fn as_slice(&self) -> &[T];
+
+    #[inline]
+    pub const fn capacity(&self) -> usize;
+
+    pub fn clear(&mut self);
+
+    pub fn dedup(&mut self);
+
+    pub fn dedup_by<F>(&mut self, same_bucket: F)
+    where
+        F: FnMut(&mut T, &mut T) -> bool;
+
+    pub fn dedup_by_key<F, K>(&mut self, key: F)
+    where
+        F: FnMut(&mut T) -> K,
+        K: PartialEq<K>;
+
+    pub fn drain<R>(&mut self, range: R)
+    where
+        R: RangeBounds<usize>;
+
+    pub fn extend_from_cloneable_slice(&mut self, other: &[T])
+    where
+        T: Clone; // Panics at run-time
+
+    pub fn extend_from_copyable_slice(&mut self, other: &[T])
+    where
+        T: Copy; // Panics at run-time
+
+    pub fn insert(&mut self, idx: usize, element: T); // Panics at run-time
+
+    #[inline]
+    pub const fn is_empty(&self) -> bool;
+
+    #[inline]
+    pub const fn len(&self) -> usize;
+
+    pub fn pop(&mut self) -> T; // Panics at run-time
+
+    pub fn push(&mut self, element: T); // Panics at run-time
```

Comment:
I quite strongly feel that the default behavior for inserts returns something akin `Result<(), InsertionError>`.

So `push` and `insert` should both return `Result`.

`pop` should also return `Option<T>`.

Reply:
> I quite strongly feel that the default behavior for inserts returns something akin Result<(), InsertionError>.


The compiler will probably eliminate the conditional bounding checking so yeah, let's do this.